ibc-clients/ics07-tendermint/src/client_state.rs

```diff
@@ -378,152 +520,269 @@ where
-                host_height,
-            )?;
-        }
-
-        Ok(vec![header_height])
+        update_client_state(self.inner(), ctx, client_id, header)
     }
 
     fn update_state_on_misbehaviour(
         &self,
         ctx: &mut E,
         client_id: &ClientId,
-        _client_message: Any,
+        client_message: Any,
     ) -> Result<(), ClientError> {
-        // NOTE: frozen height is  set to `Height {revision_height: 0,
-        // revision_number: 1}` and it is the same for all misbehaviour. This
-        // aligns with the
-        // [`ibc-go`](https://github.com/cosmos/ibc-go/blob/0e3f428e66d6fc0fc6b10d2f3c658aaa5000daf7/modules/light-clients/07-tendermint/misbehaviour.go#L18-L19)
-        // implementation.
-        let frozen_client_state = self.0.clone().with_frozen_height(Height::min(0));
-
-        let wrapped_frozen_client_state = ClientState::from(frozen_client_state);
-
-        ctx.store_client_state(
-            ClientStatePath::new(client_id.clone()),
-            wrapped_frozen_client_state.into(),
-        )?;
-
-        Ok(())
+        update_client_state_on_misbehaviour(self.inner(), ctx, client_id, client_message)
     }
 
-    // Commit the new client state and consensus state to the store
     fn update_state_on_upgrade(
         &self,
         ctx: &mut E,
         client_id: &ClientId,
         upgraded_client_state: Any,
         upgraded_consensus_state: Any,
     ) -> Result<Height, ClientError> {
-        let mut upgraded_tm_client_state = Self::try_from(upgraded_client_state)?;
-        let upgraded_tm_cons_state = TmConsensusState::try_from(upgraded_consensus_state)?;
-
-        upgraded_tm_client_state.0.zero_custom_fields();
-
-        // Construct new client state and consensus state relayer chosen client
-        // parameters are ignored. All chain-chosen parameters come from
-        // committed client, all client-chosen parameters come from current
-        // client.
-        let new_client_state = ClientStateType::new(
-            upgraded_tm_client_state.0.chain_id,
-            self.0.trust_level,
-            self.0.trusting_period,
-            upgraded_tm_client_state.0.unbonding_period,
-            self.0.max_clock_drift,
-            upgraded_tm_client_state.0.latest_height,
-            upgraded_tm_client_state.0.proof_specs,
-            upgraded_tm_client_state.0.upgrade_path,
-            self.0.allow_update,
-        )?;
+        update_client_state_on_upgrade(
+            self.inner(),
+            ctx,
+            client_id,
+            upgraded_client_state,
+            upgraded_consensus_state,
+        )
+    }
+}
+
+/// Seed the host store with initial client and consensus states.
+///
+/// Note that this function is typically implemented as part of the
+/// [`ClientStateExecution`] trait, but has been made a standalone function
+/// in order to make the ClientState APIs more flexible.
+pub fn initialise_client_state<E>(
+    client_state: &ClientState,
```

Comment:
This is the one place where the `ClientState` newtype is being passed. This is because the `ClientStateType` is missing `From<ibc_client_tendermint_types::ClientState>` for `AnyClientState`; I'm not sure that's a bound that we can satisfy, but correct me if I'm wrong.

Reply:
We can simply replace the `<E as ClientExecutionContext>::AnyClientState: From<ClientState>` constraint everywhere with `<E as ClientExecutionContext>::AnyClientState: From<ClientStateType>`, and everything should work then. This adjustment also eliminates the need for the redundant `ClientState::from(client_state)` step in situations where we need to acquire `AnyClientState` for the storing purpose.

Reply:
While at it, I think we can do so for the `AnyConsensusState` constraint too to get things consistent. 
I mean using:
    `<E as ClientExecutionContext>::AnyConsensusState: From<ConsensusStateType>`
instead of:
    `<E as ClientExecutionContext>::AnyConsensusState: From<TmConsensusState>`,

Reply:
Changing `From<TmConsensusState>` to `From<ConsensusStateType>` introduces this error:
```
error[E0277]: the trait bound `<E as ClientExecutionContext>::AnyConsensusState: core::convert::From<consensus_state::ConsensusState>` is not satisfied
   --> ibc-clients/ics07-tendermint/src/client_state.rs:767:53
    |
767 |         TmConsensusState::from(new_consensus_state).into(),
    |                                                     ^^^^ the trait `core::convert::From<consensus_state::ConsensusState>` is not implemented for `<E as ClientExecutionContext>::AnyConsensusState`
    |
    = note: required for `consensus_state::ConsensusState` to implement `core::convert::Into<<E as ClientExecutionContext>::AnyConsensusState>`
help: consider further restricting the associated type
    |
710 |     <E as ClientExecutionContext>::AnyConsensusState: From<ConsensusStateType>, <E as ClientExecutionContext>::AnyConsensusState: core::convert::From<consensus_state::ConsensusState>
```
It's not clear to me how to resolve this.

Reply:
In situations such as [here](https://github.com/cosmos/ibc-rs/blob/fa1db8c476d50e176b06f1f8a93150588ae87dc3/ibc-clients/ics07-tendermint/src/client_state.rs#L646), where we initially built the `TmConsensusState` and then converted it to the `AnyConsensusState`, the creation of `TmConsensusState` will no longer be necessary. 
Also, other places like [here](https://github.com/cosmos/ibc-rs/blob/fa1db8c476d50e176b06f1f8a93150588ae87dc3/ibc-clients/ics07-tendermint/src/client_state.rs#L571) we should obtain `tm_consensus_state` from `ConsensusStateType`.
Then, I believe you shouldn't encounter that error.